src/Analyser/MutatingScope.php

```diff
@@ -554,10 +554,13 @@ public function getVariableType(string $variableName): Type
 
 		$varExprString = '$' . $variableName;
 		if (!array_key_exists($varExprString, $this->expressionTypes)) {
-			if ($this->isGlobalVariable($variableName)) {
-				return new ArrayType(new BenevolentUnionType([new IntegerType(), new StringType()]), new MixedType(true));
+			if (!$this->isGlobalVariable($variableName)) {
+				return new MixedType();
 			}
-			return new MixedType();
+
+			$superglobalType = new ArrayType(new BenevolentUnionType([new IntegerType(), new StringType()]), new MixedType(true));
+			$this->expressionTypes[$varExprString] = ExpressionTypeHolder::createYes(new Variable($variableName), $superglobalType);
+			$this->nativeExpressionTypes[$varExprString] = ExpressionTypeHolder::createYes(new Variable($variableName), $superglobalType);
```

Comment:
I don't get why the fix looks like this.

1) This is polluting `$this->expressionTypes` with any superglobal in the analysed code.
2) It's mutating Scope in a getter method which is a big red-flag.

Reply:
good point, I don't understand how this even worked before without adding an expression tbh :P the idea of setting the expression for proper narrowing came from https://github.com/phpstan/phpstan-src/pull/3762#issuecomment-2574943147 basically.

not sure if relevant for the bug or fix, but after the expression is created it is dealing with late resolveable types still in the same method in https://github.com/phpstan/phpstan-src/blob/a11741d568f968f7b5938fb3c80e852a7952bd14/src/Analyser/MutatingScope.php#L566

Reply:
looks like I can't get around step-debugging this issue :D I assume the expressions are just overriding something that is incorrectly set and retained or so? let's see..

Reply:
I found a better fix, the root issue has to be addressed in mergeVariableHolder. Stay tuned.

Reply:
See https://github.com/phpstan/phpstan-src/pull/3901/commits/9710ba52e92e3cd45eb01af2d1761b5cfa5921d2

Reply:
oh cool, thx!

Reply:
those variable type holders were the missing piece for me I guess